### ISSUE TITLE
3.4: associations with same name

### ISSUE THREAD

**Issue by mirko-pagliai** [cakephp/cakephp#10410]
This bug occurred recently, unfortunately I had no way to immediately investigate, so I do not know from which version, maybe from 3.4.

I have `Posts` that belong to `PostsCategories` and `Pages` that belong to `PagesCategoriesTable`.

`PostsTable::initialize()`:

       /// ...
        $this->belongsTo('Categories', [
            'foreignKey' => 'category_id',
            'joinType' => 'INNER',
            'className' => 'MeCms.PostsCategories',
        ]);
       /// ...

`PagesTable::initialize()`:

       /// ...
        $this->belongsTo('Categories', [
            'foreignKey' => 'category_id',
            'joinType' => 'INNER',
            'className' => 'MeCms.PagesCategories',
        ]);
       /// ...

So, as you can see, both associations have the same name.  
I already know that this can be confusing, but until recently worked perfectly.

Now what happens?  
I have a cell that lists the post categories (with something like `$this->Posts->Categories->find()`...). Normally, this cell works.
But when the cell is rendered with an action that also uses the page categories (`$this->Pages->Categories->find()`...), this happens:

```
2017-03-13 05:41:12 - Warning (512): Could not render cell - MeCms\Model\Table\PostsTable association "Categories" of type "manyToOne" to "MeCms\Model\Table\PagesCategoriesTable" doesn't match the expected class "MeCms\Model\Table\PostsCategoriesTable". You can't have an association of the same name with a different target "className" option anywhere in your app. [/var/www/quadrifogliopsichiatria.it/vendor/cakephp/cakephp/src/ORM/Association.php, line 420] in [/var/www/quadrifogliopsichiatria.it/vendor/cakephp/cakephp/src/View/Cell.php, line 271]
```

Debugging `$this->Posts->Categories` inside the cell code there are no problems:

```
object(Cake\ORM\Association\BelongsTo) {
	[protected] _validStrategies => [
		(int) 0 => 'join',
		(int) 1 => 'select'
	]
	[protected] _name => 'Categories'
	[protected] _className => 'MeCms.PostsCategories'
	[protected] _bindingKey => null
	[protected] _foreignKey => 'category_id'
	[protected] _conditions => []
	[protected] _dependent => false
	[protected] _cascadeCallbacks => false
	[protected] _sourceTable => object(MeCms\Model\Table\PostsTable) {

		'registryAlias' => 'MeCms.Posts',
		'table' => 'posts',
		'alias' => 'Posts',
		'entityClass' => 'MeCms\Model\Entity\Post',
		'associations' => [
			(int) 0 => 'categories',
			(int) 1 => 'users',
			(int) 2 => 'tags'
		],
		'behaviors' => [
			(int) 0 => 'Timestamp',
			(int) 1 => 'CounterCache'
		],
		'defaultConnection' => 'default',
		'connectionName' => 'default'
	
	}
	[protected] _targetTable => null
	[protected] _joinType => 'INNER'
	[protected] _propertyName => null
	[protected] _strategy => 'join'
	[protected] _finder => 'all'
	[protected] _tableLocator => null
}
```
The problem occurs calling `find ()`.

The problem is [here](https://github.com/cakephp/cakephp/blob/master/src/ORM/Association.php#L396): `$this->_targetTable` is `PagesCategories`.

I have, however, noticed a peculiar thing: if specifically the class namespace, the problem does not occur. So:

```php
        $this->belongsTo('Categories', [
            'foreignKey' => 'category_id',
            'joinType' => 'INNER',
            'className' => 'MeCms\Model\Table\PostsCategoriesTable',
        ]);
```

This works.

Unfortunately I'm not to say more than that. Is it a bug? How can I help?

**Comment by inoas** [comment]
Maybe related: https://github.com/cakephp/cakephp/issues/10140#issue-204278244

**Comment by markstory** [comment]
#10142 was the pull request to add 'registryAlias' which was not merged as it lacked a real-world use case and could potentially introduce a number of undesirable side-effects.

The root of the issue here is that when a table instance is fetched from the TableLocator [the association's alias is used](https://github.com/cakephp/cakephp/blob/master/src/ORM/Association.php#L406-L411). Because you have duplicate alias names you get the 'wrong' table.

You can workaround this limitation by setting the `targetTable` option in your association definition. This will bypass the locator lookups and ensure you always get the correct instance you want.

One possible solution could be to prefix registry aliases with the current model name. Instead of looking for `Categories` in both tables, we could use `Posts__Categories` and `Pages__Categories` as the aliases. This might cause other issues though as the aliases associations are using would change. Perhaps @lorenzo has a better idea though.

**Comment by mirko-pagliai** [comment]
@markstory what do you think about my solution (using the full class name, as in the last piece of code)?

**Comment by markstory** [comment]
@mirko-pagliai I couldn't find the snippet you were trying to link to. But using the full classname as the registry alias might work as well.

**Comment by mirko-pagliai** [comment]
@markstory 

        $this->belongsTo('Categories', [
            'foreignKey' => 'category_id',
            'joinType' => 'INNER',
            'className' => 'MeCms\Model\Table\PostsCategoriesTable',
        ]);

I think it's better instead to create a table instance.

**Comment by markstory** [comment]
@mirko-pagliai The classname does let us find the class more easily, but I'm not sure it helps solve the possible conflict with the registry alias which currently is the association 'name'. I think you may have gotten lucky with load ordering when you used the full classname, as presently the classname used does not change how registry lookups are done.

**Comment by mirko-pagliai** [comment]
Surely it does not suit me to change the alias, otherwise I would change a lot of code.

I can confirm that, using the full class name, tests on tables are ok, the error disappears and the cell works:

    $this->belongsTo('Categories', [
        'foreignKey' => 'category_id',
        'joinType' => 'INNER',
        'className' => 'MeCms\Model\Table\PostsCategoriesTable',
    ]);

Instead, by using the `targetTable` option, the problem is solved, but some tests fail, for example:

        $post = $this->Posts->findById(2)->contain(['Categories'])->first();

This fails with:  
`InvalidArgumentException: You have contained 'Categories' but that association was bound as 'PostsCategories'.`

**Comment by markstory** [comment]
Interesting, that error was added to address #7110 where mismatched aliases resulted in incorrect data being added to entities.

When using the full classname, do both tables end up with the correct table bound to their `Categories` association? Based on reading the code, I don't think that is possible, as there is still only one `Categories` alias in use.

A possible solution to the registry aliases could be to not store table objects from associations in the table locator/registry, that would sidestep this issue entirely as table objects used for associations will not bleed into each other.

**Comment by mirko-pagliai** [comment]
> When using the full classname, do both tables end up with the correct table bound to their Categories association?
Yes. This evening I'll try some tests, but it seems it works.

**Comment by lorenzo** [comment]
Not registering the alias in the locator would mean looping forever building associations.

**Comment by markstory** [comment]
I chatted a bit with @lorenzo in slack, and I'm going to try and remove the error that comes up when using `targetTable` and fix any downstream errors during hydration.

**Comment by mirko-pagliai** [comment]
Tests:

    public function testA1()
    {
        $this->assertInstanceOf('Cake\ORM\Association\BelongsTo', $this->Pages->Categories);
        $this->assertEquals('Categories', $this->Pages->Categories->getName());
        $this->assertContains($this->Pages->Categories->className(), [
            'MeCms.PagesCategories',
            'MeCms\Model\Table\PagesCategoriesTable'
        ]);

        $this->assertInstanceOf('Cake\ORM\Association\BelongsTo', $this->Posts->Categories);
        $this->assertEquals('Categories', $this->Posts->Categories->getName());
        $this->assertContains($this->Posts->Categories->className(), [
            'MeCms.PostsCategories',
            'MeCms\Model\Table\PostsCategoriesTable'
        ]);

        $pageCategory = $this->Pages->Categories->find()->first();
        $this->assertInstanceof('MeCms\Model\Entity\PagesCategory', $pageCategory);

        $postsCategory = $this->Posts->Categories->find()->first();
        $this->assertInstanceof('MeCms\Model\Entity\PostsCategory', $postsCategory);
    }

    public function testA2()
    {
        $page = $this->Pages->find()->contain(['Categories'])->first();
        $this->assertInstanceOf('MeCms\Model\Entity\Page', $page);
        $this->assertInstanceof('MeCms\Model\Entity\PagesCategory', $page->category);

        $post = $this->Posts->find()->contain(['Categories'])->first();
        $this->assertInstanceOf('MeCms\Model\Entity\Post', $post);
        $this->assertInstanceof('MeCms\Model\Entity\PostsCategory', $post->category);
    }

- using `'className' => 'MeCms.PostsCategories'`:

```
There were 2 errors:

1) MeCms\Test\TestCase\View\Cell\PagesWidgetsCellTest::testA1
RuntimeException: MeCms\Model\Table\PostsTable association "Categories" of type "manyToOne" to "MeCms\Model\Table\PagesCategoriesTable" doesn't match the expected class "MeCms\Model\Table\PostsCategoriesTable". You can't have an association of the same name with a different target "className" option anywhere in your app.

/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/Association.php:420
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/Association.php:1027
/home/mirko/Libs/Plugins/MeCms/tests/TestCase/View/Cell/PagesWidgetsCellTest.php:94

2) MeCms\Test\TestCase\View\Cell\PagesWidgetsCellTest::testA2
RuntimeException: MeCms\Model\Table\PostsTable association "Categories" of type "manyToOne" to "MeCms\Model\Table\PagesCategoriesTable" doesn't match the expected class "MeCms\Model\Table\PostsCategoriesTable". You can't have an association of the same name with a different target "className" option anywhere in your app.

/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/Association.php:420
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/Query.php:413
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/Query.php:391
/home/mirko/Libs/Plugins/MeCms/tests/TestCase/View/Cell/PagesWidgetsCellTest.php:104

ERRORS!
Tests: 2, Assertions: 9, Errors: 2.
```

- using `'className' => 'MeCms\Model\Table\PostsCategoriesTable'`

```
OK (2 tests, 12 assertions)
```

- using `'targetTable' => $this->tableLocator()->get('MeCms.PostsCategories')`

```
There was 1 error:

1) MeCms\Test\TestCase\View\Cell\PagesWidgetsCellTest::testA2
InvalidArgumentException: You have contained 'Categories' but that association was bound as 'PostsCategories'.

/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/EagerLoader.php:501
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/EagerLoader.php:320
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/EagerLoader.php:450
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/EagerLoader.php:421
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/Query.php:1038
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/Query.php:989
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/Database/Driver/Mysql.php:136
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/Database/Connection.php:258
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/Database/Connection.php:311
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/Database/Query.php:213
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/Query.php:1011
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/Datasource/QueryTrait.php:276
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/ORM/Query.php:959
/home/mirko/Libs/Plugins/MeCms/vendor/cakephp/cakephp/src/Datasource/QueryTrait.php:404
/home/mirko/Libs/Plugins/MeCms/tests/TestCase/View/Cell/PagesWidgetsCellTest.php:104

--

There was 1 failure:

1) MeCms\Test\TestCase\View\Cell\PagesWidgetsCellTest::testA1
Failed asserting that an array contains null.

/home/mirko/Libs/Plugins/MeCms/tests/TestCase/View/Cell/PagesWidgetsCellTest.php:87

ERRORS!
Tests: 2, Assertions: 8, Errors: 1, Failures: 1.
```

I hope I was helpful.

**Comment by markstory** [comment]
Thank you for that. I'm not sure how the full classname is making the tests pass as its behavior shouldn't be any different than the plugin syntax. I'll keep working on fixing the errors when using `targetTable` as that will provide a fairly robust solution I think.

**Comment by mirko-pagliai** [comment]
@markstory 

> I'm not sure how the full classname is making the tests pass as its behavior shouldn't be any different than the plugin syntax

I have the same doubt. Probably it has done something wrong during the transformation from the syntax plugin to the full class name.

For now I will use the full class name, although I think this thing should be investigated.

**Comment by markstory** [comment]
Closing as I think this is resolved by #10417.
